compiler/crates/relay-compiler/src/watchman/query_builder.rs

```diff
@@ -5,22 +5,45 @@
  * LICENSE file in the root directory of this source tree.
  */
 
+use crate::compiler_state::SourceSet;
 use crate::config::{Config, SchemaLocation};
+use relay_typegen::TypegenLanguage;
 use std::path::PathBuf;
 use watchman_client::prelude::*;
 
 pub fn get_watchman_expr(config: &Config) -> Expr {
-    let mut sources_conditions = vec![
-        // ending in *.js
-        Expr::Suffix(vec!["js".into()]),
-        // in one of the source roots
-        expr_any(
-            get_source_roots(&config)
-                .into_iter()
-                .map(|path| Expr::DirName(DirNameTerm { path, depth: None }))
-                .collect(),
-        ),
-    ];
+    let mut sources_conditions = vec![expr_any(
+        config
+            .sources
+            .iter()
+            .flat_map(|(path, name)| match name {
+                SourceSet::SourceSetName(name) => {
+                    std::iter::once((path, config.projects.get(&name))).collect::<Vec<_>>()
+                }
+                SourceSet::SourceSetNames(names) => names
+                    .iter()
+                    .map(|name| (path, config.projects.get(name)))
+                    .collect::<Vec<_>>(),
+            })
+            .filter_map(|(path, project)| match project {
+                Some(p) if p.enabled => Some(Expr::All(vec![
+                    // Ending in *.js(x) or *.ts(x) depending on the project language.
+                    Expr::Suffix(match &p.typegen_config.language {
+                        TypegenLanguage::Flow => vec![PathBuf::from("js"), PathBuf::from("jsx")],
+                        TypegenLanguage::Typescript => {
+                            vec![PathBuf::from("ts"), PathBuf::from("tsx")]
```

Comment:
For a TS project, I suppose you'd have *all* your own product code in `*.tsx?` files or should this also include `.js` files for hybrid projects?

Reply:
I think generally its all ts or all js. But can't presume that, so think yeah maybe hybrids are nice?